servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java

```diff
@@ -6644,4 +6657,198 @@ public void putRoleSystemMeta(ResourceContext ctx, String domainName, String rol
+        metric.increment(ZMSConsts.HTTP_PUT);
+        logPrincipal(ctx);
+
+        if (readOnlyMode) {
+            throw ZMSUtils.requestError("Server in Maintenance Read-Only mode. Please try your request later", caller);
+        }
+
+        validateRequest(ctx.request(), caller);
+
+        validate(domainName, TYPE_DOMAIN_NAME, caller);
+        validate(roleName, TYPE_ENTITY_NAME, caller);
+        validate(memberName, TYPE_MEMBER_NAME, caller);
+        validate(membership, TYPE_MEMBERSHIP, caller);
+
+        // for consistent handling of all requests, we're going to convert
+        // all incoming object values into lower case (e.g. domain, role,
+        // policy, service, etc name)
+
+        domainName = domainName.toLowerCase();
+        roleName = roleName.toLowerCase();
+        memberName = memberName.toLowerCase();
+        AthenzObject.MEMBERSHIP.convertToLowerCase(membership);
+
+        final String principalDomain = getPrincipalDomain(ctx);
+        metric.increment(ZMSConsts.HTTP_REQUEST, domainName, principalDomain);
+        metric.increment(caller, domainName, principalDomain);
+        Object timerMetric = metric.startTiming("putmembershipdecision_timing", domainName, principalDomain);
+
+        // verify that request is properly authenticated for this request
+
+        verifyAuthorizedServiceOperation(((RsrcCtxWrapper) ctx).principal().getAuthorizedService(),
+                caller, "role", roleName);
+
+        // verify that the member name in the URI and object provided match
+
+        if (!memberName.equals(membership.getMemberName())) {
+            throw ZMSUtils.requestError("putMembershipDecision: Member name in URI and Membership object do not match", caller);
+        }
+
+        // role name is optional so we'll verify only if the value is present in the object
+
+        if (membership.getRoleName() != null && !roleName.equals(membership.getRoleName())) {
+            throw ZMSUtils.requestError("putMembershipDecision: Role name in URI and Membership object do not match", caller);
+        }
+
+        //authorization check
+        if (!isAllowedPutMembershipDecision(((RsrcCtxWrapper) ctx).principal(), domainName, roleName)) {
+            throw ZMSUtils.forbiddenError("putMembershipDecision: principal is not authorized to approve / reject members", caller);
+        }
+
+        RoleMember roleMember = new RoleMember();
+        roleMember.setMemberName(memberName);
+        roleMember.setExpiration(membership.getExpiration());
+        roleMember.setActive(membership.getActive());
+
+        dbService.executePutMembershipDecision(ctx, domainName, roleName, getNormalizedMember(roleMember), auditRef, caller);
+
+        metric.stopTiming(timerMetric, domainName, principalDomain);
+    }
+
+    private boolean isAllowedPutMembershipDecision(Principal principal, String domainName, String roleName) {
+        AthenzDomain domain = getAthenzDomain(domainName, false);
+        if (domain != null) {
```

Comment:
we shouldn't need this check. we should get the athenz domain with the option to throw exceptions if the domain is not found.